### ISSUE TITLE
Segmentation fault upon wrong named bind markers passed by driver

### ISSUE THREAD

**Issue by wprzytula** [scylladb/scylladb#12727]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 53366db6c66798808143b1e3170f2bb6dbcae5f6 (current master at the moment of writing)
Cluster size: 1
OS: Fedora

### Reproducer
Running the test `test_named_bind_markers` in Scylla Rust driver. After cloning the [repo](https://github.com/scylladb/scylla-rust-driver):
```bash
$ SCYLLA_URI=<NODE_IP> cargo test test_named_bind_markers
```

### Symptoms
A shard dies with sigsegv. Translated backtrace:
```
[Backtrace #0]
seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at reactor.cc:?
seastar::install_oneshot_signal_handler<11, (void (*)())(&seastar::sigsegv_action)>()::{lambda(int, siginfo_t*, void*)#1}::__invoke(int, siginfo_t*, void*) at reactor.cc:?
wasmtime_runtime::traphandlers::unix::trap_handler at wasmtime_runtime.2c5d786c-cgu.6:?
/lib64/libc.so.6: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=765237b0355c030ff41d969eedcb87bfccb43595, for GNU/Linux 3.2.0, not stripped

__GI___sigaction at :?
cql3::query_options::get_value_at(unsigned long) const at expression.cc:?
_ZNSt8__detail9__variant17__gen_vtable_implINS0_12_Multi_arrayIPFNS0_21__deduce_visit_resultIN4cql39raw_valueEEEO18overloaded_functorIJZNS4_4expr8evaluateERKNS8_10expressionERKNS8_17evaluation_inputsEE4$_97ZNS8_8evaluateESB_SE_E4$_98ZNS8_8evaluateESB_SE_E4$_99ZNS8_8evaluateESB_SE_E5$_100ZNS8_8evaluateESB_SE_E5$_101ZNS8_8evaluateESB_SE_E5$_102ZNS8_8evaluateESB_SE_E5$_103ZNS8_8evaluateESB_SE_E5$_104ZNS8_8evaluateESB_SE_E5$_105ZNS8_8evaluateESB_SE_E5$_106ZNS8_8evaluateESB_SE_E5$_107ZNS8_8evaluateESB_SE_E5$_108ZNS8_8evaluateESB_SE_E5$_109ZNS8_8evaluateESB_SE_E5$_110ZNS8_8evaluateESB_SE_E5$_111ZNS8_8evaluateESB_SE_E5$_112EERSt7variantIJNS8_11conjunctionENS8_15binary_operatorENS8_12column_valueENS8_5tokenENS8_21unresolved_identifierENS8_25column_mutation_attributeENS8_13function_callENS8_4castENS8_15field_selectionENS8_13bind_variableENS8_16untyped_constantENS8_8constantENS8_17tuple_constructorENS8_22collection_constructorENS8_20usertype_constructorENS8_9subscriptEEEEJEEESt16integer_sequenceImJLm9EEEE14__visit_invokeESW_S1F_ at expression.cc:?
cql3::expr::evaluate(cql3::expr::expression const&, cql3::query_options const&) at expression.cc:?
seastar::noncopyable_function<bool (cql3::expr::expression const&)>::direct_vtable_for<cql3::expr::for_each_expression<cql3::expr::binary_operator, cql3::restrictions::validate_primary_key_restrictions(cql3::query_options const&, std::vector<cql3::expr::expression, std::allocator<cql3::expr::expression> > const&)::$_15>(cql3::expr::expression const&, cql3::restrictions::validate_primary_key_restrictions(cql3::query_options const&, std::vector<cql3::expr::expression, std::allocator<cql3::expr::expression> > const&)::$_15)::{lambda(cql3::expr::expression const&)#1}>::call(seastar::noncopyable_function<bool (cql3::expr::expression const&)> const*, cql3::expr::expression const&) at statement_restrictions.cc:?
cql3::expr::recurse_until(cql3::expr::expression const&, seastar::noncopyable_function<bool (cql3::expr::expression const&)> const&) at expression.cc:?
cql3::restrictions::statement_restrictions::validate_primary_key(cql3::query_options const&) const at statement_restrictions.cc:?
cql3::statements::modification_statement::do_execute(cql3::query_processor&, service::query_state&, cql3::query_options const&) const at modification_statement.cc:?
seastar::noncopyable_function<seastar::future<seastar::shared_ptr<cql_transport::messages::result_message> > (cql3::statements::modification_statement const*, cql3::query_processor&, service::query_state&, cql3::query_options const&)>::direct_vtable_for<std::_Mem_fn<seastar::future<seastar::shared_ptr<cql_transport::messages::result_message> > (cql3::statements::modification_statement::*)(cql3::query_processor&, service::query_state&, cql3::query_options const&) const> >::call(seastar::noncopyable_function<seastar::future<seastar::shared_ptr<cql_transport::messages::result_message> > (cql3::statements::modification_statement const*, cql3::query_processor&, service::query_state&, cql3::query_options const&)> const*, cql3::statements::modification_statement const*, cql3::query_processor&, service::query_state&, cql3::query_options const&) at modification_statement.cc:?
seastar::noncopyable_function<seastar::future<seastar::shared_ptr<cql_transport::messages::result_message> > (cql3::statements::modification_statement const*, cql3::query_processor&, service::query_state&, cql3::query_options const&)>::direct_vtable_for<seastar::inheriting_concrete_execution_stage<seastar::future<seastar::shared_ptr<cql_transport::messages::result_message> >, cql3::statements::modification_statement const*, cql3::query_processor&, service::query_state&, cql3::query_options const&>::make_stage_for_group(seastar::scheduling_group)::{lambda(cql3::statements::modification_statement const*, cql3::query_processor&, service::query_state&, cql3::query_options const&)#1}>::call(seastar::noncopyable_function<seastar::future<seastar::shared_ptr<cql_transport::messages::result_message> > (cql3::statements::modification_statement const*, cql3::query_processor&, service::query_state&, cql3::query_options const&)> const*, cql3::statements::modification_statement const*, cql3::query_processor&, service::query_state&, cql3::query_options const&) at modification_statement.cc:?
seastar::concrete_execution_stage<seastar::future<seastar::shared_ptr<cql_transport::messages::result_message> >, cql3::statements::modification_statement const*, cql3::query_processor&, service::query_state&, cql3::query_options const&>::do_flush() at modification_statement.cc:?
seastar::lambda_task<seastar::execution_stage::flush()::$_1>::run_and_dispose() at execution_stage.cc:?
seastar::reactor::run_some_tasks() at reactor.cc:?
seastar::reactor::do_run() at reactor.cc:?
std::_Function_handler<void (), seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_94>::_M_invoke(std::_Any_data const&) at reactor.cc:?
seastar::posix_thread::start_routine(void*) at posix.cc:?
start_thread at ??:?
__clone3 at :?
```

**Comment by cvybhu** [comment]
I've made a minimal reproducer:
```rust
use std::collections::HashMap;

use scylla::SessionBuilder;

#[tokio::main]
async fn main() {
    let uri = std::env::var("SCYLLA_URI").unwrap_or_else(|_| "127.0.0.1:9042".to_string());
    let session = SessionBuilder::new().known_node(uri).build().await.unwrap();

    session
        .query("CREATE KEYSPACE IF NOT EXISTS named_crash WITH REPLICATION = {'class' : 'SimpleStrategy', 'replication_factor' : 1}", &[])
        .await
        .unwrap();

    session
        .query(
            "CREATE TABLE named_crash.tab (pk int, ck int, v int, PRIMARY KEY (pk, ck, v))",
            &[],
        )
        .await
        .unwrap();

    let prepared = session
        .prepare("INSERT INTO named_crash.tab (pk, ck, v) VALUES (:pk, :ck, :v)")
        .await
        .unwrap();
    let named_values: HashMap<&str, i32> =
        HashMap::from([("pk", 7), ("doesnt_exist", 42), ("ck", 13)]);
    session.execute(&prepared, &named_values).await.unwrap();
}
```
There's a problem with Scylla's validation of named bind variables. It doesn't validate that the name `doesnt_exist` does not exist and then segfaults when executing the query.

**Comment by cvybhu** [comment]
I couldn't reproduce the issue using python driver. It looks like it silently ignores being given the wrong bind variable names and chooses to set all missing values to `UNSET_VALUE`. (See [python driver documentation](https://docs.datastax.com/en/developer/python-driver/3.25/api/cassandra/query/#cassandra.query.BoundStatement-methods))

Here's a snippet I tried:
```python

import cassandra
from cassandra.cluster import Cluster
from cassandra.query import UNSET_VALUE

cql = Cluster(["127.0.0.1"]).connect()

cql.execute("DROP KEYSPACE IF EXISTS ks")
cql.execute("CREATE KEYSPACE IF NOT EXISTS ks WITH REPLICATION = {'class' : 'SimpleStrategy', 'replication_factor' : 1}")

cql.execute("CREATE TABLE ks.tab (pk int, ck int, v int, primary key (pk, ck))")
prepared = cql.prepare("INSERT INTO ks.tab (pk, ck, v) VALUES (:pk, ?, :v)")
print(prepared)

values = {'pk': 7, 'doesnt_exist': 42, 'ck': 13}

bound_stmt = prepared.bind(values)
print(bound_stmt)

cql.execute(prepared, {'pk': 7, 'doesnt_exist': 42, 'ck': 13})

rs = cql.execute("SELECT * FROM ks.tab")
for r in rs:
    print("selected row:", r)
```

This completes without any errors and prints:
```bash
<PreparedStatement query="INSERT INTO ks.tab (pk, ck, v) VALUES (:pk, ?, :v)", consistency=Not Set>
<BoundStatement query="INSERT INTO ks.tab (pk, ck, v) VALUES (:pk, ?, :v)", values=[7, 13, <object object at 0x7fb006b848d0>], consistency=Not Set>
selected row: Row(pk=7, ck=13, v=None)
```

**Comment by cvybhu** [comment]
Segfaults started happening after 0b418fa7cf164e7fcbeab2d4ab1a3412eb8aeef0. 

It made a change in `query_options::get_value_at`:
https://github.com/scylladb/scylladb/blob/61dfc9c10f1eccb4af703dc6804ddc74e9a6ba26/cql3/query_options.hh#L171-L176

Previously this method looked like this:
https://github.com/scylladb/scylladb/blob/6f055ca5f9c9dd61907d600e18b0640650040953/cql3/query_options.hh#L135-L137

The validation was lacking even before the commit that caused segfaults, but the out-of-range index used to be caught by `_value_views.at(idx);`. It threw an exception which caused an internal server error, but at least there was no crash.

**Comment by cvybhu** [comment]
The moment where `query_options` are validated is:
https://github.com/scylladb/scylladb/blob/61dfc9c10f1eccb4af703dc6804ddc74e9a6ba26/cql3/query_processor.cc#L472-L478

https://github.com/scylladb/scylladb/blob/61dfc9c10f1eccb4af703dc6804ddc74e9a6ba26/cql3/query_options.cc#L127-L146

During preparation we order the bound values by finding a value for each name and pushing them onto a vector. When there's no value with such name there should be an error, but instead no value is pushed at all.
Because of this the vector holding the bound values is too short and trying to access a bind variable with high index causes a segfault.

I will add more validation checks there.

**Comment by cvybhu** [comment]
I can't use `cql-pytest` to make a regression test :/
I will see if it can be done using boost tests.
Perhaps it might be time to finally introduce `cql-rusttest` :crab:

**Comment by cvybhu** [comment]
> I will see if it can be done using boost tests

Looks like it can't :/
`cql_test_env` only has methods which take `query_options`.
I could use them to simulate queries with named bind markers, but some validation checks happen before `query_options` are created at all.
https://github.com/scylladb/scylladb/blob/61dfc9c10f1eccb4af703dc6804ddc74e9a6ba26/test/lib/cql_test_env.hh#L99

I could move some checks to `query_options` constructor and test them manually, but those wouldn't be true end-to-end tests like `cql-pytest`

**Comment by cvybhu** [comment]
Btw we allow mixing named and non-named bind markers:
```cql
INSERT INTO ks.tab (pk, ck, v) VALUES (:pk, ?, :v)
```
This looks like a potential source of trouble, AFAIR the CQL binary protocol doesn't handle this.
I guess it deserves its own issue.

**Comment by cvybhu** [comment]
Posted a fix in #12730.

I added a `Fixes:` line, which will close this issue once the PR is merged.
I'm not sure if I should do that, this is something that might have to be backported, so the issue should maybe stay open (?).

**Comment by cvybhu** [comment]
> Btw we allow mixing named and non-named bind markers:
> 
> ```sql
> INSERT INTO ks.tab (pk, ck, v) VALUES (:pk, ?, :v)
> ```
> 
> This looks like a potential source of trouble, AFAIR the CQL binary protocol doesn't handle this. I guess it deserves its own issue.

Opened #12732

**Comment by avikivity** [comment]
Backported to 5.2 (only vulnerable branch).
